### PR TITLE
Message

### PR DESCRIPTION
**Pull request by Phenomenalhub**

### CODE REVIEW

File: app/src/main/java/com/example/studentarena/fragments/MessageFragment.java [hunk 1]
```diff
@@ -2,65 +2,131 @@
...
+        ParseQuery<Message> query2 = ParseQuery.getQuery(Message.class);
+        query2.whereEqualTo(Message.RECEIVER_KEY, (User) ParseUser.getCurrentUser());
+
+        List<ParseQuery<Message>> list = new ArrayList<ParseQuery<Message>>();
+        list.add(query1);
+        list.add(query2);
+
+        ParseQuery<Message> query = ParseQuery.or(list);
+        query.include(Message.POST_KEY);
+        MessageServices newMessageService = new MessageServices(new MessageInterface() {
```
Comment (mhammerly) [inline app/src/main/java/com/example/studentarena/fragments/MessageFragment.java:76]: i don't know how this service stuff works, pls explain to me tomorrow. i told you, you know android development better than i do :)

File: app/src/main/java/com/example/studentarena/utilities/GetRelativeTime.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    private static final int SECOND_MILLIS = 1000;
+    private static final int MINUTE_MILLIS = 60 * SECOND_MILLIS;
+    private static final int HOUR_MILLIS = 60 * MINUTE_MILLIS;
+    private static final int DAY_MILLIS = 24 * HOUR_MILLIS;
+
+    public static String getSimpleTime(Date inputDate) {
+        String twitterFormat = "EEE MMM dd HH:mm:ss ZZZZZ yyyy";
+        SimpleDateFormat sf = new SimpleDateFormat(twitterFormat, Locale.ENGLISH);
+        sf.setLenient(true);
+
```
Comment (mhammerly) [inline app/src/main/java/com/example/studentarena/utilities/GetRelativeTime.java:26]: doesn't look like you use this `sf` variable?

File: app/src/main/java/com/example/studentarena/utilities/GetRelativeTime.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+        long time = inputDate.getTime();
+        long now = System.currentTimeMillis();
```
Comment (mhammerly) [inline app/src/main/java/com/example/studentarena/utilities/GetRelativeTime.java:28]: suggested GitHub issue to follow up on later: figure out how to handle timezones

let's say you're in Washington and i'm in Michigan and we both create posts at 4:00pm. i'm three hours ahead so my post is way older, but if all parse has to go on is "4pm" it'll think we created them at the same time. to handle that, many computer things covert times to UTC (Coordinated Universal Time) before saving them. UTC is basically a special timezone that doesn't change according to Daylight Savings Time or anything, and then they convert back to a local timezone before displaying the time to the user

i think Parse automatically creates its `created_at` and `updated_at` timestamps using UTC already, but your `System.currentTimeMillis()` will get a timestamp that uses your system timezone. you probably want to convert one to the other before comparing to see how much time passed between them (see [stackoverflow](https://stackoverflow.com/questions/41062341/time-conversion-milliseconds-local-time-to-milliseconds-utc-time-in-android))

**Review by mhammerly — COMMENTED** [review]
so if i understand correctly, `ChatAdapter` / `ChatActivity` are for each item's thread with all its messages, and `MessageAdapter` / `MessageFragment` are for the list of threads with one thread per item? i have seen some apps say "ThreadView" and "ThreadList" to make clear which is which, consider adopting names like that here (or at least add comments to make it clear)

most of my comments are about tidiness, but the one about simplifying your logic for building your list of one thread per post is important

File: app/src/main/java/com/example/studentarena/ParseApplication.java [hunk 4]
```diff
@@ -2,14 +2,26 @@
...
 
     @Override
     public void onCreate() {
         super.onCreate();
+        ParseObject.registerSubclass(Message.class);
+
+        // Use for monitoring Parse OkHttp traffic
+        OkHttpClient.Builder builder = new OkHttpClient.Builder();
+        HttpLoggingInterceptor httpLoggingInterceptor = new HttpLoggingInterceptor();
+        // Can be Level.BASIC, Level.HEADERS, or Level.BODY
```
[the remark this answers is not in the document]
Comment (Phenomenalhub) [reply app/src/main/java/com/example/studentarena/ParseApplication.java:24]: Yes that was meant to be for live query to update messages

File: app/src/main/java/com/example/studentarena/fragments/ComposeFragment.java [hunk 5]
```diff
@@ -100,7 +98,7 @@ private void savePost() {
         Post post = new Post();
         post.setDescription(etDescription.getEditText().getText().toString());
         post.setContact(etContactinfo.getEditText().getText().toString());
-        post.setPrice(etPrice.getEditText().getText().toString());
+        //post.setPrice(etPrice.getEditText().getText().toString());
```
[the remark this answers is not in the document]
Comment (Phenomenalhub) [reply app/src/main/java/com/example/studentarena/fragments/ComposeFragment.java:101]: It was resolved in my search branch so I will make use of the search branch to get rid of that problem.

File: app/src/main/java/com/example/studentarena/fragments/FeedFragment.java [hunk 6]
```diff
@@ -116,6 +117,7 @@ public void done(List<Post> posts, ParseException e) {
                 // save received posts to list and notify adapter of new data
                 allPosts.addAll(posts);
                 swipeContainer.setRefreshing(false);// swipeContainer.setRefreshing(false) once the network request has completed successfully.
+                progressBar.setVisibility(View.INVISIBLE);
```
[the remark this answers is not in the document]
Comment (Phenomenalhub) [reply app/src/main/java/com/example/studentarena/fragments/FeedFragment.java:120]: This is the progress bar to show something is loading on the main feed. So it is set to invisible once it gets all post.

File: app/src/main/java/com/example/studentarena/fragments/MessageFragment.java [hunk 7]
```diff
@@ -2,65 +2,131 @@
...
+                        } else {
+                            if(m.getTargetPost()!= null){
+                                if (ParseUser.getCurrentUser().getObjectId().equals(m.getSender().getObjectId())) {
+                                postId = ParseUser.getCurrentUser().getObjectId() + " " + m.getTargetPost().getObjectId() + " " + m.getReceiver().getObjectId();
+                            } else {
+                                postId = ParseUser.getCurrentUser().getObjectId() + " " + m.getTargetPost().getObjectId() + " " + m.getSender().getObjectId();
+                            }
+                            } else{
+                                continue;
+                            }
```
[the remark this answers is not in the document]
Comment (mhammerly) [reply app/src/main/java/com/example/studentarena/fragments/MessageFragment.java:110]: we talked about the problem with my proposed code. if i sell an item, then my approach would group all interested buyers into the same thread for that item

probably the robust fix is to add a Thread model and have the detail view UI create a Thread + an initial Message in that thread. then the thread list just fetches a list of threads and no processing is needed, and when you tap on a thread it fetches just those messages and no processing is needed

say we punt on that, though. i think this logic could still be simplified:

- if target post is null
  - continue
- if current user is sender:
  - postId = senderId + postId + receiverId
- else
  - postId = receiverId + postId + senderId

**Pull request merged**